src/main/java/io/stargate/sgv2/jsonapi/service/operation/MetadataAttempt.java

```diff
@@ -0,0 +1,275 @@
+import io.stargate.sgv2.jsonapi.exception.SchemaException;
+import io.stargate.sgv2.jsonapi.service.cqldriver.executor.CommandQueryExecutor;
+import io.stargate.sgv2.jsonapi.service.cqldriver.executor.SchemaObject;
+import io.stargate.sgv2.jsonapi.service.cqldriver.executor.TableSchemaObject;
+import io.stargate.sgv2.jsonapi.service.cqldriver.executor.VectorConfig;
+import io.stargate.sgv2.jsonapi.service.schema.collections.CollectionTableMatcher;
+import io.stargate.sgv2.jsonapi.service.schema.tables.ApiDataTypeDef;
+import io.stargate.sgv2.jsonapi.service.schema.tables.ApiDataTypeDefs;
+import io.stargate.sgv2.jsonapi.util.CqlIdentifierUtil;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+
+/** An attempt to execute commands that need data from metadata */
+public abstract class MetadataAttempt<SchemaT extends SchemaObject>
+    extends OperationAttempt<MetadataAttempt<SchemaT>, SchemaT> {
+  // this will be set on executeStatement
+  private Optional<KeyspaceMetadata> keyspaceMetadata;
+
+  private static ObjectMapper objectMapper = new ObjectMapper();
+  private static final CollectionTableMatcher TABLE_MATCHER = new CollectionTableMatcher();
+
+  /**
+   * Create a new {@link OperationAttempt} with the provided position, schema object and retry
+   * policy.
+   *
+   * @param position The 0 based position of the attempt in the container of attempts. Attempts are
+   *     ordered by position, for sequential processing and for rebuilding the response in the
+   *     correct order (e.g. for inserting many documents)
+   * @param schemaObject The schema object that the operation is working with.
+   * @param retryPolicy The {@link RetryPolicy} to use when running the operation, if there is no
+   *     retry policy then use {@link RetryPolicy#NO_RETRY}
+   */
+  protected MetadataAttempt(int position, SchemaT schemaObject, RetryPolicy retryPolicy) {
+    super(position, schemaObject, retryPolicy);
+  }
+
+  protected abstract List<String> getNames();
+
+  protected abstract Object getSchema();
+
+  @Override
+  protected Uni<AsyncResultSet> executeStatement(CommandQueryExecutor queryExecutor) {
+    this.keyspaceMetadata = queryExecutor.getKeyspaceMetadata(schemaObject.name().keyspace());
+    if (keyspaceMetadata.isEmpty()) {
+      return Uni.createFrom()
+          .failure(
+              SchemaException.Code.INVALID_KEYSPACE.get(
+                  Map.of("keyspace", schemaObject.name().keyspace())));
+    }
+    return Uni.createFrom().item(new EmptyAsyncResultSet());
+  }
+
+  /**
+   * Convert table schema object to table response which is returned as response for `listTables`
+   *
+   * @return
+   */
+  protected TableResponse getTableSchema(TableSchemaObject tableSchemaObject) {
```

Comment:
if this is only use by ListTables it should not be on the super class ? 

Also created this ticket to refactor the serialisartion later so we can test it better https://github.com/stargate/data-api/issues/1544